src/Features/Core/Portable/GoToDefinition/AbstractGoToDefinitionSymbolService.cs

```diff
@@ -29,10 +31,24 @@ public async Task<(ISymbol?, TextSpan)> GetSymbolAndBoundSpanAsync(Document docu
 
             if (symbol is null)
             {
-                return (null, semanticInfo.Span);
+                return (null, project, semanticInfo.Span);
             }
 
-            return (FindRelatedExplicitlyDeclaredSymbol(symbol, semanticModel.Compilation), semanticInfo.Span);
+            // If this document is not in the primary workspace, we may want to search for results
+            // in a solution different from the one we started in. Use the starting workspace's
+            // ISymbolMappingService to get a context for searching in the proper solution.
+            var mappingService = services.GetRequiredService<ISymbolMappingService>();
+            var mapping = await mappingService.MapSymbolAsync(document, symbol, cancellationToken).ConfigureAwait(false);
```

Comment:
I'm a bit confused why we didn't have this before or why this is needed now. Find References for example does need the mapping because you want to do the search in the original workspace. Go to Definition for a classic metadata as source file wouldn't need it, since we create metadata references around the source file -- any further navigations to symbols from metadata don't need to map back since they still have good symbols there.

I'd imagine the embedded source would work the same way -- is something different here? I'm a bit more worried too if the symbol mapping fails for internal/private and whether we'll still have bugs here?

Reply:
> I'm a bit confused why we didn't have this before or why this is needed now

IMO it was a bug this wasn't done before, but MaS never hit it because it uses a project per generated document, and tears them down aggressively. Source Link is more conservative.

> since we create metadata references around the source file

We do that for Source Link too.

Reply:
Had a long conversation with @davidwengier who cleared some things up for me:

"Classic" metadata as source produces a single source file that contains a type, and we don't have partial types or anything else to worry about that way. So any go to definition from that generated file will fall into one of two categories:

1. It's a reference to another place in this same file; since that source file is "complete", it'll be fine.
2. It's a reference to a different type which we would generate into a different file. Today, since that would bind as a metadata symbol in the source file, we go back through our original process and that all works fine. And since the type is public, we generally have confidence that the type is being metadata imported by the original project.

SourceLink is a bit different in that we may have a single source file that doesn't have those same guarantees: if the type is partial, we may have only downloaded one part. But navigation within that file may require downloading other parts; navigation in a source link file can bind to a source symbol, but that still requires downloading another file. That's the case that doesn't exist in "classic" metadata as source. The idea here then being that if we can just jump back to the original file, then we restart this process from the metadata view, which is a complete view across all parts.

There's still two caveats with this approach:

1. It assumes that the source symbol can in fact be relocated in the consuming project, which may be harder if there's private members and things like that which the compiler may not import as aggressively for performance reasons.
2. It assumes that the symbol binding in the original sourcelink file is still reasonably useful. For example, consider a partial class with different overloads of the same method name in different parts. Further, there are calls in some of these source link files to overloads that live in other parts. If you have one of those open via source link, the compiler might always bind your target method for any calls to the one that lives in this part even though the overload doesn't work -- it has nothing else to return, so it just gives you that. So you can't ever F5 to the real method.

The first concern just means this fix might not always work, and we'd have to do something else in those cases which renders this change moot. The second issue isn't new to this, and we'd have to do something more aggressive generally, like always downloading via SourceLink all source files that contribute to a type whenever a navigation involved it, which has other downsides. In either case, this can't make things _worse_, it just might not be the "right" fix in that we aren't fixing the general problem somewhere else. I've asked @davidwengier to add some more comments here to this effect and explain that if we hit any of these broken cases we need to do some of the more aggressive approaches then.

Reply:
So I can just copy this wholesale into the source file, right? :)